adr/0007-package-variables-files-revisit.md

```diff
@@ -0,0 +1,135 @@
+   - *Pros*: This keeps things simpler with less to type out at the command line.
+   - *Cons*: This is less explicit and could be confusing.  It also depends on implementing a `default` pointer.
+1. Introduce a second flag (i.e. `--accept-defaults`) and error out if a user does not use the flag or set all variables through a file or at the command line
+   - *Pros*: This is more explicit and could work with the latter two `default` COAs.
+   - *Cons*: This requires another command line flag that may just get in the way.
+1. Remove the `prompt` key and detect when to prompt based on `default` existing
+   - *Pros*: This makes things simpler overall
+   - *Cons*: This relies on implementing a `default` pointer and gives package authors less control
+1. Remove the `prompt` key and always prompt
+   - *Pros*: This makes things simpler overall
+   - *Cons*: This could make things unweildy in large packages and gives package authors less control
+
+> Note that for the purposes of this ADR it was decided to exclude considering options making use of the `description` or `matches` fields as those ultimately solve different problems and would only be tangentially related to the above.
+
+## Decision
+
+It was decided to allow the `default` key to be nillable to detect its existence and error out if `--confirm` and `prompt` are used together with no existing default or `--set`.  This would result in the following behavior:
+
+```yaml
+variables:
+  - name: "NOTHING"
+    # default: nil  <- decided change
+    # prompt: false <- currently under the hood
+  - name: "NO_PROMPT"
+    default: "no-prompt"
+    # prompt: false <- currently under the hood
+  - name: "NO_DEFAULT"
+    # default: nil  <- decided change
+    prompt: true
+  - name: "BOTH"
+    default: "both"
+    prompt: true
+```
+
+```shell
+zarf package deploy package.zst --confirm --set NOTHING=hello --set NO_PROMPT=world --set NO_DEFAULT=today --set BOTH=tommorow
+# This would continue to work, setting NOTHING to "hello", NO_PROMPT to "world", NO_DEFAULT to "today" and BOTH to "tomorrow"
+
+zarf package deploy package.zst --confirm --set NOTHING=hello --set NO_DEFAULT=today
+# This would continue to work, setting NOTHING to "hello", NO_PROMPT to "no-prompt", NO_DEFAULT to "today" and BOTH to "both"
+
+zarf package deploy package.zst --confirm --set NOTHING=hello
+# This would not work, erroring out with: "Variable NO_DEFAULT does not have a default, when using `--confirm` you must specify it with `--set` or `--set-file`."
+
+zarf package deploy package.zst --confirm
+# This would not work, erroring out with: "Variable NOTHING does not have a default, when using `--confirm` you must specify it with `--set` or `--set-file`."
+
+zarf package deploy package.zst
+# This would not work, erroring out with: "Variable NOTHING does not have a default, and is not prompted for interactively. You must specify it with `--set` or `--set-file`."
```

Comment:
I'm not really getting how this case is valuable to the user. It would make more sense to me for Zarf to just go ahead and prompt for this variable if the user is in "interactive mode".

I see `prompt: true` as valuable only when paired with a default, in cases where the package author wants to provide a suggestion while also giving the user of the package 
![image](https://user-images.githubusercontent.com/16000938/186714065-b43da950-489e-4527-8818-801d54d07883.png)

Reply:
hang on that diagram is a big wrong. fixing it now..

Reply:
![image](https://user-images.githubusercontent.com/16000938/186715964-e8f63448-ae29-4456-9b68-dc02c237a830.png)

Reply:
Here is more what I was thinking from the ADR (this is also close to current behavior except there is no "Default -> No" step)
![image](https://user-images.githubusercontent.com/3977569/186716605-f0ff595d-b2e4-41df-afd4-d367b1aee4e4.png)

Reply:
ah yep, you're right that it would not cause a full cycle of the feedback loop if they can --set it on the CLI. As long as the error message says that (which is happening in the current proposal), then that's likely fine, though it would require them to use a flag when they were previously not using any

Reply:
I like the idea of changing `prompt:` which defaults to false to `suppressPrompt:` which defaults to false. Then it is an intentional action by the package author to let the user get down to the Fail at the bottom, rather than an unintended consequence of forgetting to add the field.

Reply:
And here is the full logic tree: (@Madeline-UX for context as well)

![image](https://user-images.githubusercontent.com/3977569/186718313-cd1bd3dc-56f0-427e-a02e-066924d0fc97.png)

Reply:
And for more clarity, here is the currently implemented workflow:

![image](https://user-images.githubusercontent.com/3977569/186721074-2fc5ca3d-c265-452e-a43e-b8762b027335.png)

Reply:
@YrrepNoj thoughts on the above and potential changing of `prompt` to `suppressPrompt` (or similar)?

Reply:
I like both the logic tree and the change of `prompt` -> `suppressPrompt`

Reply:
A breaking change to make asking uses in interactive mode the default behavior seems logical to me, we could make it a little shorter with noPrompt as well.

Reply:
I think this captures the decision point between "prompt" and "noPrompt". The benefit of "noPrompt" is that the failure condition is pulled out of the "path of least resistance" in the case where both package author and user don't know or care about taking additional action to avoid failure.
![image](https://user-images.githubusercontent.com/16000938/186736821-987b4676-7b52-4c22-bfd6-ab234308a413.png)

Reply:
Yeah I like the idea of switching to `noPrompt` and will update the ADR accordingly since it looks like we all agree.

Reply:
ADR updated here: https://github.com/defenseunicorns/zarf/blob/675-load-package-variables-from-file/adr/0007-package-variables-files-revisit.md

Reply:
Another way to handle prompting. 

I recommend that we move error validation to the package create. 

Ex. Upon package create if Zarf identifies any defaults that are left empty, it returns a message to the package creator reading something like this and requiring them to confirm.

Zarf detected these variables are missing a default value. If you do not specify a default value a value of "empty string" will be automatically set as the default. 

- [values and missing default] [Add value or edit value button]- [values and missing default] [Add value or edit value button]

Reply:
Updated the ADR to remove this on package create: https://github.com/defenseunicorns/zarf/blob/675-load-package-variables-from-file/adr/0007-package-variables-files-revisit.md